### ISSUE TITLE
Feature Request / Ideas

### ISSUE THREAD

**Issue by cmprmsd** [nesfit/fitcrack#21]
Hey!
I was now able to test the solution and compared to hashtopolis e.g. it has a much better UI feeling. I don't know about performance yet, but you paper states it's advantages in this regard as well.

Some observations I made, or features that I miss:
- [ ] exporting cracked passwords
- [ ] hashlists with usernames
- [ ] rerun a job with only hashes that have not been cracked yet and with only new (selectable) wordlists or rulesets.
- [ ] pass multiple rulesets to Hashcat (Hashcat will combine them to apply all rules of the second set on the first set). This will especially benefit cracking complex passwords that are based on known keywords like the company name. I really like the fact that you can create password lists from the web ui. 

The width of the password file UI is quite small. I think that's a displaying bug, that I could fix on the client side.

**Comment by ihranicky** [comment]
Hello there, thank you for feedback on our system. The features you mention seem great for improving the user experience. We will consider including them in the future updates.

For the performance - it is very much about how you configure and tweak the system, concretely:
- For a job, the **Desired time per workunit** (aka chunk size in Hashtopolis) plays a bit role.
- While Hashtopolis uses its own protocol, Fitcrack utilizes BOINC. So on the host, I recommend checking BOINC Manager -> Options -> Computing preferences: Here you can specify (or disable) host limitations, e.g. Suspended when non-boinc CPU usage is above xx%, etc.
- To boost-up Hashcat's performance, you can create ``/etc/fitcrack.conf`` (Linux) or ``C:\ProgramData\BOINC\fitcrack.conf`` on a host machine and set up additional parameters. Whatever you type there will be passed as an extra parameter to hashcat. For instance, if you type ``-w 4`` it will switch the workload profile to Nightmare (highest performance).
- For advanced tweaking, you can play with the values in Webadmin -> System -> Settings -> System preferences.

.

**Comment by cmprmsd** [comment]
That's a good hint with fitcrack.conf 👍
Adding -O also adds a good boost to hashcat, when your passwords are not too long.

I'll extend this list with other stuff, I notice while using it. 
Great to see, that you maintain this project!

Btw, as workaround for reruns with new dicts, I'd copy the passwords into a new passwordlist starting with `_a` and select it with the new pwlist in order to immediately crack them again and skip the attempts for the rest of the lists and rules. 
However, tracking the hashes before passing them to new jobs or reruns would be an ideal solution.

**Comment by Allesanddro** [comment]
Would also like to have the ability to export cracked passwords

**Comment by davidbolvansky** [comment]
> Would also like to have the ability to export cracked passwords

Implemented in dev branch.

![image](https://user-images.githubusercontent.com/22410393/165980579-410f1e56-9543-4362-8930-3a71e3273278.png)

**Comment by thegnomedesu** [comment]
The ability to give clients extra commands such as -w or -O from the web interface would be nice. The ability to leverage CPU on PRINCE Attacks would be nice too, from the small amount of reading ive done PRINCE does well on CPU compared to GPU.

**Comment by davidbolvansky** [comment]
I think we add -O always. But yeah, I agree that this would be useful.

**Comment by davidbolvansky** [comment]
> What about assigning jobs to CPU? This would be a nice option for certain attack types.

Should not be hard to implement.

**Comment by lachlan2k** [comment]
> I think we add -O always. But yeah, I agree that this would be useful.

`-O` reduces the maximum length of password that can be cracked. On some algorithms this is quite significant (down to 16 characters). Is there a way to disable it?

**Comment by davidbolvansky** [comment]
Now, after https://github.com/nesfit/fitcrack/commit/d36285f7ceddb822095ef55339427d8abfab48f0, yes.

Please try.

**Comment by Pipkill-prog** [comment]
Whats about add the ability to remove or edit Hashes/Passwords on the Libary?

**Comment by davidbolvansky** [comment]
You mean something like “Delete” button to remove “items” (hashes) in Library -> Hashes?

Can you describe more what do you mean by “Edit” ? Remove found password for certain hash? Please share more details what you would like to see here.

Thanks for the feedback!

**Comment by Pipkill-prog** [comment]
> You mean something like “Delete” button to remove “items” (hashes) in Library -> Hashes?

Yes, sometimes it can happen that inexperienced Users insert wrong formatet Hashes (NT-Hash with LM Part oder straight NTDS.dit files) and it would be nice if you can delete/edit those wrong Hashes

> Can you describe more what do you mean by “Edit” ? Remove found password for certain hash? Please share more details what you would like to see here.

Yes, something like this or repair a Password when its displayed as Hexfile because the Password contains foreign Chars

Thanks for your fast Answer ;)

**Comment by davidbolvansky** [comment]
> Also being able to add new dictionaries into exhausted jobs would be incredibly useful

You can select multiple dictionaries for the job, not only one. So I dont think there is a need for this suggest feature.. or?

**Comment by davidbolvansky** [comment]
> The ability to give clients extra commands such as -w or -O from the web interface would be nice. The ability to leverage CPU on PRINCE Attacks would be nice too, from the small amount of reading ive done PRINCE does well on CPU compared to GPU.

Now done in dev branch.

![image](https://github.com/nesfit/fitcrack/assets/22410393/0e6684b5-d851-42c9-a096-181cd955b552)

**Comment by davidbolvansky** [comment]
> You mean something like “Delete” button to remove “items” (hashes) in Library -> Hashes?
> 
> Can you describe more what do you mean by “Edit” ? Remove found password for certain hash? Please share more details what you would like to see here.
> 
> Thanks for the feedback!

Just observed how slow "Hashes" page is with thousands of hashes. This needs to be fixed as well.

**Comment by opengs** [comment]
> > The ability to give clients extra commands such as -w or -O from the web interface would be nice. The ability to leverage CPU on PRINCE Attacks would be nice too, from the small amount of reading ive done PRINCE does well on CPU compared to GPU.
> 
> Now done in dev branch.
> 
> ![image](https://private-user-images.githubusercontent.com/22410393/247478980-0e6684b5-d851-42c9-a096-181cd955b552.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzMTA4MjQsIm5iZiI6MTcxNDMxMDUyNCwicGF0aCI6Ii8yMjQxMDM5My8yNDc0Nzg5ODAtMGU2Njg0YjUtZDg1MS00MmM5LWEwOTYtMTgxY2Q5NTViNTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDEzMjIwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5MDZlYjZiMTY5MDBiMDA5MDk5ZGY5ZmUxZWVjODUwN2UyYjZjYjUyZjRiMTQzZWIzOWJhMmEzZWE2NGM5ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aRui0FwAVGPxMXB2RbX33uBbjsGk6q1ysPvTg6syn1U)

@davidbolvansky any possibility that those changes will be pushed to `dev` or to `master`?

**Comment by burgbosh** [comment]
PCFG is a very good feature to expand the existing password database, but it is very slow for multiple GPUs. You can upgrade to the latest updates. https://github.com/lakiw /, and add functions that allow you to assign custom tasks to each GPU by running a separate copy of pcfg-manager
